lib/components/prometheusmetrics.js

```diff
@@ -0,0 +1,315 @@
+"use strict";
+
+var fs = require("fs");
+
+var TICKS_PER_SEC = 100; // TODO(paul): look this up via sysconf(_SC_CLK_TCK)
+
+/**
+ * Prometheus-style /metrics gathering and exporting.
+ * This class provides a central location to register gauge and counter metrics
+ * used to generate the <code>/metrics</code> page.
+ *
+ * It also contains collector code to generate the standard
+ * <code>process_</code> and <code>nodejs_</code> metrics, which depends on a
+ * working <code>/proc</code> filesystem and therefore is only active on Linux.
+ * The standard metrics exported in this case are:
+ *   process_resident_memory_bytes
+ *   process_virtual_memory_bytes
+ *   process_heap_bytes
+ *   process_cpu_seconds_total
+ *   process_open_fds
+ *   process_max_fds
+ *   process_start_time_seconds
+ *   nodejs_heap_used_bytes
+ *
+ * This class depends on having <code>prometheus-client</code> installed. It
+ * will attempt to load this module when the constructor is invoked.
+ *
+ * @example <caption>A simple metric that counts the keys in an object:</caption>
+ *   var metrics = new PrometheusMetrics();
+ *
+ *   var orange = {};
+ *   metrics.addGauge({
+ *       name: "oranges",
+ *       help: "current number of oranges",
+ *       refresh: (gauge) => {
+ *           gauge.set({}, Object.keys(oranges).length);
+ *       },
+ *   });
+ *
+ * @example <caption>Generating values for multiple gauges in a single collector function.</caption>
+ *   var metrics = new PrometheusMetrics();
+ *
+ *   var oranges_gauge = metrics.addGauge({
+ *       name: "oranges",
+ *       help: "current number of oranges",
+ *   });
+ *   var apples_gauge = metrics.addGauge({
+ *       name: "apples",
+ *       help: "current number of apples",
+ *   });
+ *
+ *   metrics.addCollector(() => {
+ *       var counts = this._countFruit();
+ *       oranges_gauge.set({}, counts.oranges);
+ *       apples_gauge.set({}, counts.apples);
+ *   });
+ *
+ * @example <caption>Using counters</caption>
+ *   var metrics = new PrometheusMetrics();
+ *
+ *   metrics.addCollector({
+ *       name: "things_made",
+ *       help: "count of things that we have made",
+ *   });
+ *
+ *   function makeThing() {
+ *       metrics.incCounter("things_made");
+ *       return new Thing();
+ *   }
+ *
+ * @constructor
+ */
+function PrometheusMetrics() {
+    // Only attempt to load these dependencies if metrics are enabled
+    var Prometheus = require("prometheus-client");
+
+    var client = this._client = new Prometheus();
+
+    this._collectors = []; // executed in order
+    this._counters = {}; // counter metrics keyed by name
+
+    var have_proc_stat = false;
+    try {
+        fs.readFileSync("/proc/self/stat");
```

Comment:
> Whlie this is a Sync form of call, using the async form won't actually make any difference. The /proc/self/ introspection API in the kernel is implemented in an interesting way that doesn't play nicely with aio, so libuv's async versions of file reading won't be any different on those.
In any case, the /proc entries won't block on reading; they're designed to be fast and lightweight to respond

Please can you put some version of said comment here to make it clear for any future readers.